src/block/Hopper.php

```diff
@@ -92,8 +99,231 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper){
+			return;
+		}
+
+		$transferCooldown = $tile->getTransferCooldown();
+		if($transferCooldown > 0){
+			$transferCooldown--;
+			$tile->setTransferCooldown($transferCooldown);
+		}
+
+		if($this->isPowered() || $transferCooldown > 0){
+			return;
+		}
+
+		$inventory = $tile->getInventory();
+		$success = $this->push($inventory);
+		// Hoppers that have a container above them, won't try to pick up items.
+		$origin = $this->position->getWorld()->getTile($this->position->getSide(Facing::UP));
+		//TODO: Not all blocks a hopper can pull from have an inventory (for example: Jukebox).
+		if($origin instanceof Container){
+			$success = $this->pull($inventory, $origin->getInventory()) || $success;
+		}else{
+			$success = $this->pickup($inventory) || $success;
```

Comment:
To avoid retrieving entities from above when a block is obstructing, it would be interesting to check whether the block above is full, and therefore that no entity can be above it.

Reply:
You still have to check whether an item entity is inside that full block:
`It is also possible for a hopper to collect items from inside a full, solid block, a situation that might come from items rising up through solid blocks or being summoned.` [wiki](https://minecraft.fandom.com/wiki/Hopper#Collecting_items)

Reply:
This very specific case may not need to be taken into account in pocketmine. It may increase the performance impact of hoppers, when in fact you should just place a full block on top to prevent entities from being checked.  

And this is what some java engines use to optimize performance.

Reply:
Of course, it's an increased performance impact. Blocking this is basically just disabling a - mostly pretty useless - feature, making PM less-matching with vanilla behaviour.
In the end, it's your decision on what you want to implement into the core. But imo, before removing functionality for performance's sake, there should be taken a look at other, more important hot paths.

Reply:
I agree that pocketmine wouldn't exactly follow vanilla's behavior in this situation, but is that so important?
It's something to think about, as the feature is not very useful, but it could greatly improve performance as soon as the number of hoppers becomes too large.

Reply:
To me it doesn't seem like a situation that would come up very often anyway, so the performance gain from changing it is likely to be close to zero. I'd classify this as premature optimisation.

Reply:
I think @ShockedPlot7560's suggestion was, to skip the entire pickup logic (calling getNearbyEntities(), etc.) if there is a full block above the hopper, or even don't reschedule the hopper if he also has nowhere to push items too.
That should actually help and not just be premature optimization. But as I said, it's not really an optimization, since it removes vanilla behaviour.